### PR TITLE
Adapt ServerCall to respect the Proxy system properties.

### PR DESCRIPTION
**Pull request by mfriedenhagen**
ServerCall should respect the Proxy system properties.
For more information, see https://docs.oracle.com/javase/8/docs/technotes/guides/net/proxies.html

### CODE REVIEW

File: agent-common/test/com/thoughtworks/go/agent/launcher/ProxyConfiguratorTest.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    }
+
+    @Test
+    public void shouldReturnProxyHost() {
+        assertNotNull(ProxyConfigurator.create(method, "proxy.example.org", "3128", null));
+    }
+
+    @Test
+    public void shouldReturnProxyHostWhenHostDoesNotMatchNonProxyHosts() {
+        assertNotNull(ProxyConfigurator.create(method, "proxy.example.org", "3128", "www.example.org"));
```
Comment (ketan) [inline agent-common/test/com/thoughtworks/go/agent/launcher/ProxyConfiguratorTest.java]: I'd expect that the assertion happen on the state of the `ProxyHost` instance that is returned, not just that an object exists — 

```
assertThat(proxyHost.getPort(), is(...))
assertThat(proxyHost.getHost(), is(...))
assertThat(proxyHost.getProtocol(), is(...))
```

File: agent-common/src/com/thoughtworks/go/agent/launcher/ServerCall.java [hunk 2]
```diff
@@ -45,6 +45,11 @@ public static ServerResponseWrapper invoke(HttpMethod method) throws Exception {
         HashMap<String, String> headers = new HashMap<String, String>();
         HttpClient httpClient = new HttpClient();
         httpClient.setConnectionTimeout(HTTP_TIMEOUT_IN_MILLISECONDS);
+        final String proxyHost = System.getProperty("http.proxyHost");
+        final String proxyPort = System.getProperty("http.proxyPort");
+        final String nonProxyHosts = System.getProperty("http.nonProxyHosts");
```
Comment (ketan) [inline agent-common/src/com/thoughtworks/go/agent/launcher/ServerCall.java]: Would you consider adding support for `https.proxyHost` and `https.proxyPort` as well?

**Comment by ketan** [comment]
I see that you've already signed the CLA. If you can address the few pieces of things I've asked, this PR will be good to merge :+1:

**Comment by mfriedenhagen** [comment]
Hello @ketan, the proxy configuration is only used for `ServerCall` right now. I would guess, that at least `com.thoughtworks.go.agent.launcher.ServerBinaryDownloader#download` should take the proxy into account as well.

**Comment by mfriedenhagen** [comment]
IntelliJ reports the following code fragments where HttpClient is used as well:
```
New instance creation
agent
com.thoughtworks.go.agent.functional
X509CertificateTest
shouldSaveCertificateInAgentTrustStore()
        HttpClient client = new HttpClient();
com.thoughtworks.go.agent.service
SslInfrastructureService
SslInfrastructureServiceTest
agent-common
com.thoughtworks.go.agent.launcher
ServerBinaryDownloader
download()
        HttpClient httpClient = new HttpClient();
ServerCall
common
com.thoughtworks.go.util
HttpService
HttpService()
        this(new HttpClientFactory(new HttpClient()));
server
com.thoughtworks.go.helpers
HttpClientHelper
doRequest(String, RequestMethod, String)
        HttpClient client = new HttpClient();
com.thoughtworks.go.util
TimeReportingUtil
report()
        HttpClient httpClient = new HttpClient();
```

**Comment by ketan** [comment]
@mfriedenhagen — We're close to a 16.3 release later this week (or early next). Please give me a few days to circle back to you on this.

**Comment by vivekjain10** [comment]
Thanks @mfriedenhagen for taking this up :+1:

**Comment by mfriedenhagen** [comment]
I guess the proxy settings should probably be taken into account for every HttpClient. What would be the appropriate Maven module, then?

**Comment by vivekjain10** [comment]
@ketan any possibility of getting this change in?

**Comment by ketan** [comment]
Closing in favor of #2236. I've upgraded httpclient to 4.5.2 in that PR so it should be easier to configure it in one place.

**Pull request closed without merging**